File: src/02.kt
import aoc.common.getParsedLines

fun solve(commands: List<Command>): Pair<Int, Int> {
    var hpos = 0
    var depth_1 = 0
    var depth_2 = 0
    var aim = 0
    for (command in commands) {
        when (command.direction) {
            "forward" -> {
                hpos += command.value
                depth_2 += aim * command.value
            }
            "up" -> {
                depth_1 -= command.value
                aim -= command.value
            }
            "down" -> {
                depth_1 += command.value
                aim += command.value
            }
        }
    }
    return Pair(hpos * depth_1, hpos * depth_2)
}

data class Command(val direction: String, val value: Int)

fun parseCommand(line: String): Command {
    val parts = line.split(" ")
    return Command(parts[0], parts[1].toInt())
}

fun main(args: Array<String>) {
    val commands = getParsedLines<Command>(args[0], ::parseCommand)
    val (result1, result2) = solve(commands)
    println(result1)
    println(result2)
}

File: src/02.erl
-module(aoc).
-export([main/1]).
-mode(compile).

-include("common/common.hrl").

solve(Commands) ->
    solve(Commands, 0, 0, 0, 0).

solve([], HPos, Depth1, Depth2, _) ->
    {results, HPos * Depth1, HPos * Depth2};

solve([{command, Direction, Value} | Rest], HPos, Depth1, Depth2, Aim) ->
    case Direction of
        forward -> solve(Rest, HPos+Value, Depth1, Depth2+Aim*Value, Aim);
        up -> solve(Rest, HPos, Depth1-Value, Depth2, Aim-Value);
        down -> solve(Rest, HPos, Depth1+Value, Depth2, Aim+Value)
    end.

parse_command(Line) ->
    [Part1, Part2 | _] = string:tokens(Line, [$\s]),
    {command, list_to_atom(Part1), list_to_integer(Part2)}.

main([Filename]) ->
    Commands = get_parsed_lines(Filename, fun parse_command/1),
    {results, Result1, Result2} = solve(Commands),
    io:fwrite("~w~n~w~n", [Result1, Result2]).
